polkadot/runtime/common/src/slots/mod.rs

```diff
@@ -310,27 +384,112 @@ impl<T: Config> Pallet<T> {
 	// Useful when trying to clean up a parachain leases, as this would tell
 	// you all the balances you need to unreserve.
 	fn all_deposits_held(para: ParaId) -> Vec<(T::AccountId, BalanceOf<T>)> {
-		let mut tracker = sp_std::collections::btree_map::BTreeMap::new();
-		Leases::<T>::get(para).into_iter().for_each(|lease| match lease {
-			Some((who, amount)) => match tracker.get(&who) {
-				Some(prev_amount) =>
-					if amount > *prev_amount {
-						tracker.insert(who, amount);
+		ReservedAmounts::<T>::iter_prefix(para).collect::<Vec<_>>()
+	}
+
+	/// Returns how much deposit should be taken from the leaser for the current lease period of the
+	/// parachain. This is the maximum of all upcoming leases for the parachain with the same
+	/// leaser.
+	fn required_deposit(para: ParaId, leaser: &T::AccountId) -> BalanceOf<T> {
+		Leases::<T>::get(para)
+			.into_iter()
+			.map(|lease| match lease {
+				Some((who, amount)) =>
+					if &who == leaser {
+						amount
+					} else {
+						Zero::zero()
 					},
-				None => {
-					tracker.insert(who, amount);
-				},
-			},
-			None => {},
+				None => Zero::zero(),
+			})
+			.max()
+			.unwrap_or_else(Zero::zero)
+	}
+
+	/// Returns true if the current lease period is ending within the next `T::EarliestRefundPeriod`
+	/// blocks.
+	fn lease_period_ending_soon(now: BlockNumberFor<T>) -> bool {
+		if let Some(current_lease) = Self::lease_period_index(now).map(|(index, _)| index) {
+			let soon = now.saturating_add(T::EarliestRefundPeriod::get());
+
+			// default to current lease period if we can't find the lease period index.
+			let soon_lease =
+				Self::lease_period_index(soon).map(|(index, _)| index).unwrap_or(current_lease);
+
+			return current_lease
+				.checked_add(&sp_runtime::traits::One::one())
+				.map(|next_lease| soon_lease == next_lease)
+				.unwrap_or(false)
+		}
+
+		false
+	}
+
+	/// Reserve the amount for the parachain, updating the current reserved amount for this leaser
+	/// and parachain.
+	fn reserve(para: ParaId, leaser: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+		T::Currency::reserve(&leaser, amount).map(|_| {
+			ReservedAmounts::<T>::mutate(para, leaser, |maybe_current| {
+				if let Some(current) = maybe_current {
+					*current = amount.saturating_add(*current)
+				} else {
+					*maybe_current = Some(amount)
+				}
+			});
+			()
+		})
```

Comment:
Maybe a dumb Rust question, but won't the `|_|` swallow an error and just pass it into the closure? As in, if `leaser` didn't have enough funds to reserve, then it would still return `Ok(())`?

Reply:
`map` only applies on OK value. `map_err` is used for transforming err values.
https://doc.rust-lang.org/std/result/enum.Result.html#method.map